src/node.js

```diff
@@ -390,6 +391,78 @@
     }
   };
 
+  startup.processPromises = function() {
+    // this should be kept in sync with PromiseRejectEvent in v8.h
+    var kPromiseRejectWithNoHandler = 0;
+    var kPromiseHandlerAddedAfterReject = 1;
+
+    var timers = NativeModule.require('timers');
+    var unhandledPromises = new WeakMap();
+    // A hack to work around uniterable WeakMaps.
+    var unhandledPromiseKeys = [];
+    var fullTurnQueued = false;
+
+    function queueFullTurn() {
+      if (!fullTurnQueued) {
+        fullTurnQueued = true;
+        timers.setImmediate(function() {
+          process.nextTick(afterFullTurn);
```

Comment:
Good question- it really boils down to the distinction between a microtask (like nextTick) and a macrotask (setImmediate) - a setTimeout would be arbitrary- what we want to do is to wait for IO to flush and the event loop to loop (hence the setImmediate) but we also want to wait for other subscribers that started a macrotask at the same time we did so a microtask is scheduled. I think @Domenic is probably better than me at explaining these delicacies but that's the general gist of things. It's certainly not random or arbitrary :)